### PR TITLE
added explain for AGNews Bert classification example

### PR DESCRIPTION
**Pull request by Arvind-Ideas2IT**
Signed-off-by: Arvind-Ideas2IT <arvindkumarsingh.gautam@ideas2it.com>

## What changes are proposed in this pull request?

Added Explain method for AGNews classification Bert Example

This PR is dependent on - https://github.com/mlflow/mlflow-torchserve/pull/70- Adding Explain cli option for mlflow torchserve plugin

## How is this patch tested?
 Ran  locally and all the check got passed
(Details)

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow TorchServe Deployment Plugin users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s) does this PR affect?
Components
- [ ] `area/deploy`: Main deployment plugin logic
- [ ] `area/build`: Build and test infrastructure for MLflow TorchServe Deployment Plugin
- [ ] `area/docs`: MLflow TorchServe Deployment Plugin documentation pages
- [x ] `area/examples`: Example code

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: examples/E2EBert/news_classifier.py [hunk 1]
```diff
@@ -57,22 +58,24 @@ def __getitem__(self, item):
...
             add_special_tokens=True,
-            max_length=self.max_length,
+            max_length=100,
             return_token_type_ids=False,
             padding="max_length",
             return_attention_mask=True,
             return_tensors="pt",
             truncation=True,
         )
+        # input_ids = torch.tensor([self.tokenizer.encode(
```
Comment (shrinath-suresh) [inline examples/E2EBert/news_classifier.py:71]: Can you please remove the commented lines

File: examples/E2EBert/news_classifier.py [hunk 2]
```diff
@@ -300,9 +339,10 @@ def training_step(self, train_batch, batch_idx):
         :return: output - Training loss
         """
         input_ids = train_batch["input_ids"]
-        attention_mask = train_batch["attention_mask"]
+        input_ids = input_ids.to(device)
```
Comment (shrinath-suresh) [inline examples/E2EBert/news_classifier.py:342]: Can you please check if `.to(device)` is needed when using PTL

File: examples/E2EBert/news_classifier.py [hunk 3]
```diff
@@ -300,9 +339,10 @@ def training_step(self, train_batch, batch_idx):
...
         targets = train_batch["targets"]
-        output = self.forward(input_ids, attention_mask)
+        targets = targets.to(device)
```
Comment (shrinath-suresh) [inline examples/E2EBert/news_classifier.py:344]: Can you please check if `.to(device)` is needed when using PTL

File: examples/E2EBert/news_classifier_handler.py [hunk 4]
```diff
@@ -74,37 +76,25 @@ def preprocess(self, data):
...
         :return: output - predicted output
         """
-
-        self.model.eval()
-        inputs = encoding.to(self.device)
-        outputs = self.model.forward(**inputs)
-
-        out = np.argmax(outputs.cpu().detach())
-        return [out.item()]
+        inputs = self.input_ids.to(self.device)
```
Comment (shrinath-suresh) [inline examples/E2EBert/news_classifier_handler.py:94]: same here

File: examples/E2EBert/news_classifier_handler.py [hunk 5]
```diff
@@ -121,3 +111,80 @@ def postprocess(self, inference_output):
...
+        model_wrapper = AGNewsmodelWrapper(self.model)
+        tokenizer = BertTokenizer(self.VOCAB_FILE)
+        model_wrapper.eval()
+        model_wrapper.zero_grad()
+        input_ids = torch.tensor([tokenizer.encode(self.text, add_special_tokens=True)])
+        input_embedding_test = model_wrapper.model.bert_model.embeddings(input_ids)
+        preds = model_wrapper(input_embedding_test)
+        out = np.argmax(preds.cpu().detach(), axis=1)
+        out = out.item()
+        ig_1 = IntegratedGradients(model_wrapper)
```
Comment (shrinath-suresh) [inline examples/E2EBert/news_classifier_handler.py:177]:
```suggestion
```

Comment (shrinath-suresh) [reply examples/E2EBert/news_classifier_handler.py:177]: `self.ig` can be used

**Review by shrinath-suresh — CHANGES_REQUESTED** [review]

File: examples/E2EBert/news_classifier.py [hunk 6]
```diff
@@ -57,22 +58,24 @@ def __getitem__(self, item):
         """
         review = str(self.reviews[item])
         target = self.targets[item]
-
-        encoding = self.tokenizer.encode_plus(
+        encoded = self.tokenizer.encode_plus(
             review,
             add_special_tokens=True,
-            max_length=self.max_length,
```
[the remark this answers is not in the document]
Comment (Arvind-Ideas2IT) [reply examples/E2EBert/news_classifier.py:64]: @shrinath-suresh  No, Actually as you know, when we were not getting the right embedding, I was trying with encoder instead of encode_plus. I am already working on updating this var as it was before now.

**Comment by shrinath-suresh** [comment]
@Arvind-Ideas2IT Can you please update the README.md file as well ?

File: examples/E2EBert/news_classifier.py (hunk 1 above)
Comment (Arvind-Ideas2IT) [reply examples/E2EBert/news_classifier.py:71]: @shrinath-suresh  Sure.

File: examples/E2EBert/news_classifier.py (hunk 2 above)
Comment (Arvind-Ideas2IT) [reply examples/E2EBert/news_classifier.py:342]: @shrinath-suresh  I belive we need this anyway I will cross check

File: examples/E2EBert/news_classifier_handler.py (hunk 5 above)
Comment (Arvind-Ideas2IT) [reply examples/E2EBert/news_classifier_handler.py:175]: @shrinath-suresh I have addressed that

**Review by shrinath-suresh — APPROVED** [review]
LGTM

File: examples/E2EBert/news_classifier.py [hunk 7]
```diff
@@ -23,6 +23,8 @@
 import torchtext.datasets as td
 from transformers import BertModel, BertTokenizer, AdamW
 
+device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
```
Comment (HamidShojanazeri) [inline examples/E2EBert/news_classifier.py:26]: @Arvind-Ideas2IT please use cuda instead of cuda with gpu id and remove the commented code on Line#431

File: examples/E2EBert/news_classifier_handler.py [hunk 8]
```diff
@@ -121,3 +109,80 @@ def postprocess(self, inference_output):
...
+            data_preprocess (Torch Tensor): Preprocessed data to be used for captum
+            raw_data (list): The unprocessed data to get target from the request
+        Returns:
+            dict : A dictionary response with the explanations response."""
+        vis_data_records_base = []
+        model_wrapper = AGNewsmodelWrapper(self.model)
+        tokenizer = BertTokenizer(self.VOCAB_FILE)
+        model_wrapper.eval()
+        model_wrapper.zero_grad()
+        input_ids = torch.tensor([tokenizer.encode(self.text, add_special_tokens=True)])
```
Comment (HamidShojanazeri) [inline examples/E2EBert/news_classifier_handler.py:170]: @Arvind-Ideas2IT Please move input_ids to device, following it needs changes in Lines #126 and #144/ please run a test on gpu if possible.

File: examples/E2EBert/README.md [hunk 9]
```diff
@@ -87,21 +88,22 @@ Create an empty directory `model_store` and run the following command to start T
...
 
 
-## Running prediction based on deployed model
+## Running prediction and explain based on deployed model
 
 For testing the fine tuned model, a sample input text is placed in `input.json`
 Run the following command to invoke prediction of our sample input 
 
 `mlflow deployments predict --name news_classification_test --target torchserve --input-path input.json  --output-path output.json`
+`mlflow deployments explain --name news_classification_test --target torchserve --input-path input.json  --output-path output.json`
```
Comment (HamidShojanazeri) [inline examples/E2EBert/README.md:107]: @Arvind-Ideas2IT Please check the formats in the md what shows up, there no space between these two command.

**Review by HamidShojanazeri — CHANGES_REQUESTED** [review]

File: examples/E2EBert/news_classifier.py (hunk 7 above)
Comment (Arvind-Ideas2IT) [reply examples/E2EBert/news_classifier.py]: @HamidShojanazeri  Thanks for pointing it out.  I have updated this.

File: examples/E2EBert/news_classifier_handler.py (hunk 8 above)
Comment (Arvind-Ideas2IT) [reply examples/E2EBert/news_classifier_handler.py]: @HamidShojanazeri  I have updated this as well and made changes in line #126 & #144 as to use  attributions.cpu().detach().numpy()  and np.argmax(output.cpu().detach()) respectively

File: examples/E2EBert/README.md (hunk 9 above)
Comment (Arvind-Ideas2IT) [reply examples/E2EBert/README.md:112]: @HamidShojanazeri Done.

**Comment by shrinath-suresh** [comment]
@HamidShojanazeri Sorry to bump the thread. Looks like @Arvind-Ideas2IT has addressed the review comments. if you are okay with the changes, shall we go ahead and merge the PR?

File: examples/E2EBert/news_classifier.py [hunk 10]
```diff
@@ -23,6 +23,8 @@
 import torchtext.datasets as td
 from transformers import BertModel, BertTokenizer, AdamW
 
+device = torch.device("cuda:id" if torch.cuda.is_available() else "cpu")
```
Comment (HamidShojanazeri) [inline examples/E2EBert/news_classifier.py:26]: @Arvind-Ideas2IT This is not able to pick up cuda:id, please change it to cuda.

**Review by HamidShojanazeri — COMMENTED** [review]
@Arvind-Ideas2IT what is the torchtext version requirement for it? is it 0.9.0?

**Comment by shrinath-suresh** [comment]
[bert-captum-explain.txt](https://github.com/mlflow/mlflow-torchserve/files/6765956/bert-captum-explain.txt)

Fixed the example, handler and tested create/predict and explain. 

This PR is dependent on #70 and mlflow needs to be compiled from source (master - as it contains mlflow explain api changes). We can merge #70 first and test it once again.

**Comment by shrinath-suresh** [comment]
@HamidShojanazeri The dependent PR #70 has been merged and i have retested the example in both cpu and gpu. Please find the logs attached.

[bert-captum-gpu.txt](https://github.com/mlflow/mlflow-torchserve/files/6801036/bert-captum-gpu.txt)
[bert-captum.txt](https://github.com/mlflow/mlflow-torchserve/files/6801037/bert-captum.txt)

Here are the pointers for testing the example:

Install mlflow from source - `pip install git+https://github.com/mlflow/mlflow.git`
Install mlflow-torchserve from source - `pip install git+https://github.com/mlflow/mlflow-torchserve.git`

These 2 changes are needed, as the explain cli command is available in source (plugins are yet to be released).

if testing the ddp scenario- make sure to install torchmetrics - `pip install torchmetrics==0.4.1`  - There is a model saving issue in ddp which is fixed in this branch - https://github.com/PyTorchLightning/pytorch-lightning/issues/8227

All the deployment commands are same as in README.md file. Added here for quick reference.

Training - 

CPU: `python news_classifier.py --max_epochs 1 --num_samples 100`
GPU: `python news_classifier.py --max_epochs 1 --num_samples 100 --gpus 2 --accelerator ddp`

Create deployment: `mlflow deployments create -t torchserve -m state_dict.pth --name news_classification_test -C "MODEL_FILE=news_classifier.py" -C "HANDLER=news_classifier_handler.py" -C "EXTRA_FILES=class_mapping.json,bert_base_uncased_vocab.txt,wrapper.py"`

predict: `mlflow deployments predict --name news_classification_test --target torchserve --input-path input.json`

explain: `mlflow deployments explain --name news_classification_test --target torchserve --input-path input.json`

Please review and let me know your comments.

**Review by HamidShojanazeri — APPROVED** [review]
Thanks @shrinath-suresh, could successfully test on my end.

**Review by chauhang — APPROVED** [review]
